src/librustc_mir/borrow_check/nll/type_check/mod.rs

```diff
@@ -1534,16 +1534,17 @@ impl<'a, 'gcx, 'tcx> TypeChecker<'a, 'gcx, 'tcx> {
         predicates: impl IntoIterator<Item = ty::Predicate<'tcx>>,
         location: Location,
     ) {
-        let mut predicates_iter = predicates.into_iter();
+        let predicates_vec = predicates.into_iter().by_ref().collect::<Vec<_>>();
```

Comment:
yikes, this seems excessive. why not just duplicate the iterator?

Reply:
that was the first thing I've tried but I can't ...

```
error[E0599]: no method named `clone` found for type `impl IntoIterator<Item = ty::Predicate<'tcx>>` in the current scope
```

Reply:
We were talking with @pnkfelix to maybe do that thing only in the debug case and avoid the extra cost when debug is off. He said that it may not worth, I'm ok with changing this to whatever is better :).